src/cursor/cur_file.c

```diff
@@ -122,9 +122,14 @@ __curfile_compare(WT_CURSOR *a, WT_CURSOR *b, int *cmpp)
     WT_CURSOR_BTREE *cbt;
     WT_DECL_RET;
     WT_SESSION_IMPL *session;
+    bool tracked;
```

Comment:
The pattern here is pretty ugly, in terms of using a tracked variable and repeating that a bunch of times in the file.

I started trying to use an API macro, but gave up due to nuances and complexity in the tracking macros that I didn't want to make worse. Then I realized that sometimes the API tracking macros go to error directly, so this tracking could be skipped, and thus the local variable to know what has happened.
I thought about adding a field to the WT_CURSOR structure, and passing that to the track_cursor API to avoid the local variable, but I got worried that it would be hard to ensure it is reset in all cases..

I'm open to better ideas.

Reply:
I think it's worth questioning the premise of not tracking calls that end up failing the setup macro. The stats count threads in API calls, and I don't see why a failed API call setup shouldn't count towards that. 

Theoretically (but hopefully not in practice) the setup macro could take a long time, or get stuck in an infinite loop. And I'd definitely want to see how many threads were in API calls if that happened.

My head isn't as deep into this space as yours, but it appears that would let us drop the `tracked` variables and the explicit `__wt_api_track_cursor_(start|end)` (instead bundling them into the macros).

Reply:
I had started there (by nesting this inside the other API tracking macros), but it became too complex.

It's hard to move this outside of the API entry macros - those are supposed to be the first and last things an API does (I'm sure there is at least one exception, but generally). The functionality here also relies on api call counters to track recursion that is tracked via the macros.

It isn't common for a macro to hang before returning an error. For example, I noticed this because some of our Python tests deliberately try to do illegal operations on prepared transactions - failures of which are flagged via the macro.

Reply:
Thanks - I would like to take you up on the Zoom meeting, I think there's something I'm missing about the problems that having the call in the macro causes.

The thing I don't get here is: we have a set of macros that are called at the beginning and end of each interesting function. Even if they jump directly to `err`, all uses call the right `API_END` macro. So I'd have thought that putting the `__wt_api_track_cursor_X` calls towards the beginning of the ~5 relevant `CURSOR_API*` macros, before any failures are possible, would "naturally" result in the correct counting behaviour.